core/src/graph/graph.rs

```diff
@@ -248,52 +249,95 @@ impl Graph {
 
 		// Now try to add new connections into pages already being updated
 		// Note: these pages have already been cloned, so we don't clone them again
-		let mut add_iter = ids_to_add.iter();
-		while let Some(_) = add_iter.clone().peekable().peek() {
-			if let Some((_page_id, page)) =
-				updated_pages.iter_mut().find(|(_, page)| !self.is_page_full(&page, false))
-			{
-				let id_to_add = add_iter.next().unwrap(); // safe to unwrap because we peeked above
-				page.add_connection(id_to_add)?;
-			} else {
+		let mut add_iter = ids_to_add.iter().cloned().peekable();
+		for aggressive in vec![false, true] {
+			for page in updated_pages.values_mut() {
+				self.add_to_page_until_full(page, &mut add_iter, aggressive, dsnp_version_config);
```

Comment:
nit: this would still iterate pages even though the `add_iter` might be finished